tools/lint/cmd_feature.c

```diff
@@ -27,65 +27,74 @@
 void
 cmd_feature_help(void)
 {
-    printf("Usage: feature [-h] <module> [<module>]*\n"
-            "                  Print features of all the module with state of each one.\n");
-}
-
-int
-collect_features(const struct lys_module *mod, struct ly_set *set)
-{
-    struct lysp_feature *f = NULL;
-    uint32_t idx = 0;
-
-    while ((f = lysp_feature_next(f, mod->parsed, &idx))) {
-        if (ly_set_add(set, (void *)f->name, 1, NULL)) {
-            YLMSG_E("Memory allocation failed.\n");
-            ly_set_erase(set, NULL);
-            return 1;
-        }
-    }
-
-    return 0;
+    printf("Usage: feature [-g] <module> [<module>]* | [-a]\n"
+            "                  Print features of all the modules with state of each one.\n\n"
+            "   -g <module1, module2, ...>, --generate <module1, module2, ...>\n"
+            "                   Generate features parameter for the command \"add\" "
+                                "in the form of -F <module-name>:<features>\n"
+            "   -a \n"
+            "                   Print features of all implemented modules.\n");
 }
 
 void
 cmd_feature(struct ly_ctx **ctx, const char *cmdline)
 {
     int argc = 0;
     char **argv = NULL;
+    char *features_output = NULL;
     int opt, opt_index, i;
+    ly_bool generate_features = 0, print_all = 0;
+    struct ly_set set = {0};
+    const struct lys_module *mod;
     struct option options[] = {
         {"help", no_argument, NULL, 'h'},
+        {"all", no_argument, NULL, 'a'},
+        {"generate", no_argument, NULL, 'g'},
         {NULL, 0, NULL, 0}
     };
-    struct ly_set set = {0};
-    size_t max_len;
-    uint32_t j;
-    const char *name;
 
     if (parse_cmdline(cmdline, &argc, &argv)) {
         goto cleanup;
     }
 
-    while ((opt = getopt_long(argc, argv, "h", options, &opt_index)) != -1) {
+    while ((opt = getopt_long(argc, argv, "hag", options, &opt_index)) != -1) {
         switch (opt) {
         case 'h':
             cmd_feature_help();
             goto cleanup;
+        case 'a':
+            print_all = 1;
+            break;
+        case 'g':
+            generate_features = 1;
+            break;
         default:
             YLMSG_E("Unknown option.\n");
             goto cleanup;
         }
     }
 
+    if (print_all) {
+        if (print_all_features(*ctx, generate_features, &features_output)) {
+            YLMSG_E("Printing all features failed.\n");
+            goto cleanup;
+        }
+        if (generate_features) {
+            printf("%s\n", features_output);
```

Comment:
So, if you run `feature <module>`, the module's features will get printed and their status (on/off).
If you run `feature -a`, it will do the same as above, but for all the implemented modules.
If you run `feature -g <module>`, a parameter generated from the given module's features will be printed (e.g., `feature -g ietf-tls-common` -> ` -F ietf-tls-common:tls10,tls11,tls12,tls13,hello-params,public-key-generation`).
Lastly, if you run `feature -g -a`, it will be a combination of the second and third case, meaning all the implemented modules' features will be printed, along with their status, and at the end the generated parameter.

Hence I chose `Usage: feature [-g] <module> [<module>]* | [-a]`.

Reply:
Ok, thanks for the clarification. I would choose:
`Usage: feature [-g] <module> [<module>]*`
`...... feature -a [-g]`  (replace the dots with spaces)

to avoid confusion, if the operator `|` refers to `<module> [<module>]*` or `[-g] <module> [<module>]*`